les/serverpool.go

```diff
@@ -17,904 +17,283 @@
-		return
-	}
-	for _, e := range list {
-		log.Debug("Loaded server stats", "id", e.node.ID(), "fails", e.lastConnected.fails,
-			"conn", fmt.Sprintf("%v/%v", e.connectStats.avg, e.connectStats.weight),
-			"delay", fmt.Sprintf("%v/%v", time.Duration(e.delayStats.avg), e.delayStats.weight),
-			"response", fmt.Sprintf("%v/%v", time.Duration(e.responseStats.avg), e.responseStats.weight),
-			"timeout", fmt.Sprintf("%v/%v", e.timeoutStats.avg, e.timeoutStats.weight))
-		pool.entries[e.node.ID()] = e
-		if pool.trustedNodes[e.node.ID()] == nil {
-			pool.knownQueue.setLatest(e)
-			pool.knownSelect.Update((*knownEntry)(e))
+	for _, url := range trustedURLs {
+		if node, err := enode.Parse(validSchemes, url); err == nil {
+			s.trusted = append(s.trusted, node.ID())
+		} else {
+			log.Error("Invalid trusted server URL", "url", url, "error", err)
 		}
 	}
-}
 
-// connectToTrustedNodes adds trusted server nodes as static trusted peers.
-//
-// Note: trusted nodes are not handled by the server pool logic, they are not
-// added to either the known or new selection pools. They are connected/reconnected
-// by p2p.Server whenever possible.
-func (pool *serverPool) connectToTrustedNodes() {
-	//connect to trusted nodes
-	for _, node := range pool.trustedNodes {
-		pool.server.AddTrustedPeer(node)
-		pool.server.AddPeer(node)
-		log.Debug("Added trusted node", "id", node.ID().String())
+	s.mixer = enode.NewFairMix(mixerTimeout)
+	knownSelector := lpc.NewWrsIterator(s.ns, s.stHasValue, s.ns.StatesMask(disableSelection), sfSelected, sfiEnr, s.knownSelectWeight, validSchemes)
+	alwaysConnect := lpc.NewQueueIterator(s.ns, s.stAlwaysConnect, s.ns.StatesMask(disableSelection), sfSelected, sfiEnr, validSchemes)
+	s.mixSources = append(s.mixSources, knownSelector)
+	s.mixSources = append(s.mixSources, alwaysConnect)
+	if discovery != nil {
+		discEnrStored := enode.Filter(discovery, func(node *enode.Node) bool {
+			s.ns.UpdateState(node.ID(), stDiscovered, 0, time.Hour)
+			s.ns.SetField(node.ID(), s.enrFieldId, node.Record())
+			return true
+		})
+		s.mixSources = append(s.mixSources, discEnrStored)
 	}
-}
 
-// parseTrustedNodes returns valid and parsed enodes
-func parseTrustedNodes(trustedNodes []string) map[enode.ID]*enode.Node {
-	nodes := make(map[enode.ID]*enode.Node)
+	// preNegotiationFilter will be added in series with iter here when les4 is available
 
-	for _, node := range trustedNodes {
-		node, err := enode.Parse(enode.ValidSchemes, node)
-		if err != nil {
-			log.Warn("Trusted node URL invalid", "enode", node, "err", err)
-			continue
+	s.dialIterator = enode.Filter(s.mixer, func(node *enode.Node) bool {
+		n, _ := s.ns.GetField(node.ID(), s.nodeHistoryFieldId).(*nodeHistory)
+		if n == nil {
+			n = &nodeHistory{}
+			s.ns.SetField(node.ID(), s.nodeHistoryFieldId, n)
 		}
-		nodes[node.ID()] = node
-	}
-	return nodes
+		n.lock.Lock()
+		n.dialCost.Add(dialCost, s.vt.StatsExpirer().LogOffset(s.clock.Now()))
```

Comment:
let's make it a member function of `nodeHistory`. So that we don't need to do `LOCK` `UNLOCK` opts.

Reply:
Now `nodeHistory` is stored in its field as a value instead of a reference so we don't need locking any more.